crates/executor/src/execution_state.rs

```diff
@@ -23,21 +33,25 @@ pub struct ExecutionState {
 impl ExecutionState {
     pub(super) fn starknet_state(
         &mut self,
-    ) -> anyhow::Result<(CachedState<PathfinderStateReader<'_>>, BlockContext)> {
+    ) -> anyhow::Result<(
+        CachedState<LruCachedReader<PathfinderStateReader<'_>>>,
+        BlockContext,
+    )> {
         let block_context = super::block_context::construct_block_context(self)?;
 
-        let state_reader = PathfinderStateReader::new(
+        let raw_reader = PathfinderStateReader::new(
             &mut self.connection,
             self.state_at_block,
             self.pending_update.is_some(),
         )?;
-
-        let mut state = CachedState::new(state_reader, GlobalContractCache::default());
+        let lru_cached_reader = LruCachedReader::new(CONTRACT_CACHE.clone(), raw_reader)?;
+        // The `CachedState` is actually more than a simple cache. It doesn't overlap with the LRU cache.
+        let mut cached_state = CachedState::new(lru_cached_reader, GlobalContractCache::default());
```

Comment:
Isn't using our `CONTRACT_CACHE` here equivalent to the `LruCachedReader` wrapper?

Reply:
The `LruCachedReader` implements `StateReader` and contains the fallback+insert mechanism, while the `CONTRACT_CACHE` is a pure, dumb cache.
Maybe I could move the latter with the former and simplify the constructor.

Reply:
Yeah, but if you check `CachedState` internals it turns out that it already implements caching contract loads in a global cache (second argument to the constructor):

https://github.com/starkware-libs/blockifier/blob/19e54f74243afa72069d440eab799d13f0dbf352/crates/blockifier/src/state/cached_state.rs#L226

I think we're missing some calls to `move_classes_to_global_cache` though: https://github.com/starkware-libs/blockifier/blob/19e54f74243afa72069d440eab799d13f0dbf352/crates/blockifier/src/state/cached_state.rs#L100

Reply:
I slightly refactored my original solution to share both caches, but I couldn't find an easy and elegant way to leverage the `move_classes_to_global_cache` method.
Part of the problem here is that some methods exposed by `blockifier` and that we use don't take a generic `State` as a parameter but its `CachedState` implementation, so there is no way to use a wrapping approach similar to the one I've used.
All in all, I expect relying `move_classes_to_global_cache` would give us similar results anyway.